test/libsolidity/util/ContractABIUtils.cpp

```diff
@@ -33,55 +39,256 @@ using namespace dev::solidity::test;
+						addressTypeParams.push_back(Parameter{bytes(), "", type, FormatInfo{}});
+					for (auto const& type: valueTypes)
+						valueTypeParams.push_back(Parameter{bytes(), "", type, FormatInfo{}});
+					for (auto const& type: dynamicTypes)
+						dynamicTypeParams.push_back(Parameter{bytes(), "", type, FormatInfo{}});
+				}
+				else
+				{
+					_errorReporter.warning(
+						"Could not convert \"" + type +
+						"\" to internal ABI type representation. Falling back to default encoding."
+					);
+					return boost::none;
+				}
+
+				finalParams += addressTypeParams + valueTypeParams;
+
+				addressTypeParams.clear();
+				valueTypeParams.clear();
 			}
+	}
 
-	return abiParams;
+	return boost::optional<ParameterList>(finalParams + dynamicTypeParams);
 }
 
-std::vector<ABIType> ContractABIUtils::fromTypeName(string const& _type) const
+bool ContractABIUtils::appendTypesFromName(
+	Json::Value const& _functionOutput,
+	ABITypes& _addressTypes,
+	ABITypes& _valueTypes,
+	ABITypes& _dynamicTypes,
+	bool _isCompoundType
+)
 {
-	static regex s_boolType{"(bool)"};
-	static regex s_uintType{"(uint\\d*)"};
-	static regex s_intType{"(int\\d*)"};
-	static regex s_bytesType{"(bytes\\d+)"};
-	static regex s_dynBytesType{"(\\bbytes\\b)"};
-	static regex s_stringType{"(string)"};
+	string type = _functionOutput["type"].asString();
+	if (isBool(type))
+		_valueTypes.push_back(ABIType{ABIType::Boolean});
+	else if (isUint(type))
+		_valueTypes.push_back(ABIType{ABIType::UnsignedDec});
+	else if (isInt(type))
+		_valueTypes.push_back(ABIType{ABIType::SignedDec});
+	else if (isBytes(type))
+		_valueTypes.push_back(ABIType{ABIType::Hex});
+	else if (isString(type))
+	{
+		_addressTypes.push_back(ABIType{ABIType::Hex});
 
-	vector<ABIType> abiTypes;
-	if (regex_match(_type, s_boolType))
-		abiTypes.push_back(ABIType{ABIType::Boolean, ABIType::AlignRight, 32});
-	else if (regex_match(_type, s_uintType))
-		abiTypes.push_back(ABIType{ABIType::UnsignedDec, ABIType::AlignRight, 32});
-	else if (regex_match(_type, s_intType))
-		abiTypes.push_back(ABIType{ABIType::SignedDec, ABIType::AlignRight, 32});
-	else if (regex_match(_type, s_bytesType))
-		abiTypes.push_back(ABIType{ABIType::Hex, ABIType::AlignRight, 32});
-	else if (regex_match(_type, s_dynBytesType))
+		if (_isCompoundType)
+			_dynamicTypes.push_back(ABIType{ABIType::Hex});
+
+		_dynamicTypes.push_back(ABIType{ABIType::UnsignedDec});
+		_dynamicTypes.push_back(ABIType{ABIType::String, ABIType::AlignLeft});
+	}
+	else if (isTuple(type))
 	{
-		abiTypes.push_back(ABIType{ABIType::UnsignedDec, ABIType::AlignRight, 32});
-		abiTypes.push_back(ABIType{ABIType::UnsignedDec, ABIType::AlignRight, 32});
-		abiTypes.push_back(ABIType{ABIType::HexString, ABIType::AlignLeft, 32});
+		for (auto const& component: _functionOutput["components"])
+			appendTypesFromName(component, _addressTypes, _valueTypes, _dynamicTypes, true);
 	}
-	else if (regex_match(_type, s_stringType))
+	else if (isBoolArray(type))
```

Comment:
What about dynamic arrays?

Reply:
They are not supported yet. We would need to read the actual array size returned by the call in order to generate enough ABI types. But here, we are falling back to a decimal representation for the uncovered, but will change it to hex with: https://github.com/ethereum/solidity/pull/7091.